### DISCUSSION TITLE
Support for JSON functions with mixed type/arity like json_build_object

### DISCUSSION THREAD

**Discussion by kelleyvanevert** [oguimbal/pg-mem#151] [Ideas]
First of all: I **love** 🎉 that this library exists, and want to use it super much :)

But, in my project, I use a bunch of JSON stuff. I tried adding `json_build_object` via `db.public.registerFunction`, but can't get it to work due to its variable arity and typing. I'm thinking, maybe the overload-checking of `registerFunction` could be (optionally) inverted? E.g. instead of setting the `args` option, setting an option like `acceptArgs: (args) => boolean`? ..and then also passing the types along with the args when invoking the function somehow.

Has this idea, or the JSON functions, been floated before? I could put it some work into a fork, if this proposal or something similar sounds reasonable?

(Another issue I'm running into, btw, is using `table.*`, which I use a lot as in `json_agg(table.*)`.)

**Comment by oguimbal** [comment]
We could think of something like that, yes. That makes sense.
Thanks for considering a contribution !

That said, I think it would be more like `acceptArgs:  (args: IType[]) => { returns: IType, args: IType[]}`  to give the chance to the end user to tell which kind of argument are actually accepted (so the runner can try to convert arguments before calling the types), and which is the return type of this overload.

For instance:

- If you declared `jsonb_set` the way you describe, the runner would call `acceptArgs` with types `string, string, string` when executing `select jsonb_set('{"a":42}', '{b}', '55')`, but `jsonb, string, string` when executing `select jsonb_set('{"a":42}'::jsonb, '{b}', '55')`.
👉 The actual implementation of this function would receive either serialized strings, or plain objects, depending on the performed query.

- If you do it the way I suggest, it should fix the problem:
   1) The `acceptArgs` will be given `string, string, string`, and will return `{returns: json, args: [json, text, json]}`
   2) The runner will force-convert the given arguments to the argument types specified by `acceptArgs`
   3)  The fn implementation will receive correct types :) 

As for the json_agg, that's another topic ! I suggest we discuss it on another thread 🙂

Sorry for the late reply.
